packages/kbn-health-gateway-server/src/kibana/routes/status.ts

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import https from 'https';
+import type { Request, ResponseToolkit } from '@hapi/hapi';
+import nodeFetch, { RequestInit, Response } from 'node-fetch';
+import type { IConfigService } from '@kbn/config';
+import type { Logger } from '@kbn/logging';
+import type { KibanaConfigType } from '../kibana_config';
+import { KibanaConfig } from '../kibana_config';
+
+const GATEWAY_STATUS_ROUTE = '/api/status';
+const KIBANA_STATUS_ROUTE = '/api/status';
+
+interface StatusRouteDependencies {
+  log: Logger;
+  config: IConfigService;
+}
+
+type Fetch = (path: string) => Promise<Response>;
+
+export function createStatusRoute({ config, log }: StatusRouteDependencies) {
+  const kibanaConfig = new KibanaConfig(config.atPathSync<KibanaConfigType>('kibana'));
+  const fetch = configureFetch(kibanaConfig);
+
+  return {
+    method: 'GET',
+    path: GATEWAY_STATUS_ROUTE,
+    handler: async (req: Request, h: ResponseToolkit) => {
+      const responses = await fetchKibanaStatuses({ fetch, kibanaConfig, log });
+      const { body, statusCode } = mergeStatusResponses(responses);
+      return h.response(body).type('application/json').code(statusCode);
+    },
+  };
+}
+
+async function fetchKibanaStatuses({
+  fetch,
+  kibanaConfig,
+  log,
+}: {
+  fetch: Fetch;
+  kibanaConfig: KibanaConfig;
+  log: Logger;
+}) {
+  const requests = await Promise.allSettled(
+    kibanaConfig.hosts.map(async (host) => {
+      log.debug(`Fetching response from ${host}${KIBANA_STATUS_ROUTE}`);
+      const response = await fetch(`${host}${KIBANA_STATUS_ROUTE}`);
+      const responseJson = await response.json();
+      log.info(
+        `Got response from ${host}${KIBANA_STATUS_ROUTE}: ${JSON.stringify(
+          responseJson.status.overall
+        )}`
+      );
+      return response;
+    })
+  );
+
+  return requests.map((r, i) => {
+    if (r.status === 'fulfilled') {
+      return r.value;
+    } else {
+      const message = `Unable to retrieve status from [${kibanaConfig.hosts[i]}]: ${JSON.stringify(
+        r
+      )}`;
+      log.error(message);
+      throw new Error(message);
```

Comment:
I know it's on the todo list, but integration tests would make it a bit easier to understand the behaviour. 

But I think throwing here means the route handler will return a generic 500 response. I think we should reserver 500 for "something bad happened that we didn't expect". Whereas a host not responding is an expected condition and even if that host isn't online, there's nothing wrong with this health gateway.

It ties into the other todo of the aggregation strategy. But whatever the final format I think we should always provide as much context as possible. E.g. all the hosts and their statuses or in this case that we had an error determining the status of a specific host (and probably status code 503).

So I would not throw here but rather map to `message`. Although it's useful to have the complete error in the logs, the message we return should probably not have the error in it (unecessary disclosure).

Reply:
> Although it's useful to have the complete error in the logs, the message we return should probably not have the error in it (unecessary disclosure).

That's a good point, for now I went ahead and pulled out the full error, instead simply checking if the responses were fulfilled or rejected, returning a 503 if there was a failure. We will still need to do a more detailed follow-up to handle different error scenarios, though.